File: app/src/main/java/org/kiwix/kiwixmobile/data/DataSource.java
/*
 * Kiwix Android
 * Copyright (c) 2019 Kiwix <android.kiwix.org>
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 */

package org.kiwix.kiwixmobile.data;

import io.reactivex.Completable;
import io.reactivex.Flowable;
import io.reactivex.Single;
import java.util.List;
import org.kiwix.kiwixmobile.bookmark.BookmarkItem;
import org.kiwix.kiwixmobile.history.HistoryListItem;
import org.kiwix.kiwixmobile.zim_manager.Language;
import org.kiwix.kiwixmobile.zim_manager.fileselect_view.adapter.BooksOnDiskListItem;

/**
 * Defines the set of methods which are required to provide the presenter with the requisite data.
 */

public interface DataSource {
  Single<List<BooksOnDiskListItem>> getLanguageCategorizedBooks();

  Completable saveBook(BooksOnDiskListItem.BookOnDisk book);

  Completable saveBooks(List<BooksOnDiskListItem.BookOnDisk> book);

  Completable saveLanguages(List<Language> languages);

  Single<List<HistoryListItem>> getDateCategorizedHistory(boolean showHistoryCurrentBook);

  Completable saveHistory(HistoryListItem.HistoryItem history);

  Completable deleteHistory(List<HistoryListItem> historyList);

  Completable clearHistory();

  Single<List<BookmarkItem>> getBookmarks(boolean showFromCurrentBook);

  Single<List<String>> getCurrentZimBookmarksUrl();

  Completable saveBookmark(BookmarkItem bookmark);

  Completable deleteBookmarks(List<BookmarkItem> bookmarks);

  Completable deleteBookmark(String bookmarkUrl);

  Flowable<List<BooksOnDiskListItem>> booksOnDiskAsListItems();
}

File: core/src/main/java/org/kiwix/kiwixmobile/core/data/DataSource.kt
/*
 * Kiwix Android
 * Copyright (c) 2019 Kiwix <android.kiwix.org>
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 */
package org.kiwix.kiwixmobile.core.data

import kotlinx.coroutines.flow.Flow
import org.kiwix.kiwixmobile.core.dao.entities.WebViewHistoryEntity
import org.kiwix.kiwixmobile.core.page.bookmark.adapter.LibkiwixBookmarkItem
import org.kiwix.kiwixmobile.core.page.history.adapter.HistoryListItem
import org.kiwix.kiwixmobile.core.page.history.adapter.HistoryListItem.HistoryItem
import org.kiwix.kiwixmobile.core.page.notes.adapter.NoteListItem
import org.kiwix.kiwixmobile.core.zim_manager.fileselect_view.BooksOnDiskListItem
import org.kiwix.libkiwix.Book

/**
 * Defines the set of methods which are required to provide the presenter with the requisite data.
 */
interface DataSource {
  fun getLanguageCategorizedBooks(): Flow<List<BooksOnDiskListItem>>
  suspend fun saveBook(book: Book)
  suspend fun saveBooks(book: List<Book>)
  suspend fun saveHistory(history: HistoryItem)
  suspend fun deleteHistory(historyList: List<HistoryListItem>)
  suspend fun clearHistory()
  fun getBookmarks(): Flow<List<LibkiwixBookmarkItem>>
  suspend fun getCurrentZimBookmarksUrl(): List<String>
  suspend fun saveBookmark(libkiwixBookmarkItem: LibkiwixBookmarkItem)
  suspend fun deleteBookmarks(bookmarks: List<LibkiwixBookmarkItem>)
  suspend fun deleteBookmark(bookId: String, bookmarkUrl: String)
  fun booksOnDiskAsListItems(): Flow<List<BooksOnDiskListItem>>
  suspend fun saveNote(noteListItem: NoteListItem)
  suspend fun deleteNote(noteTitle: String)
  suspend fun deleteNotes(noteList: List<NoteListItem>)
  suspend fun insertWebViewPageHistoryItems(webViewHistoryEntityList: List<WebViewHistoryEntity>)
  fun getAllWebViewPagesHistory(): Flow<List<WebViewHistoryEntity>>
  suspend fun clearWebViewPagesHistory()
}
